File: python/sumRootToLeaf.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


def sumNumbers(root: TreeNode | None) -> int:
    sumPaths = [0]
    traverse(root, "", sumPaths)
    return sumPaths[0]


def traverse(root: TreeNode | None, path: str, pathsSum: list[int]):
    if root is None:
        return

    path += str(root.val)
    if root.left is None and root.right is None:
        pathsSum[0] += int(path)
        return

    traverse(root.left, path, pathsSum)
    traverse(root.right, path, pathsSum)


left = TreeNode(val=2)
right = TreeNode(val=3)
root = TreeNode(val=1, left=left, right=right)
print(sumNumbers(root))

File: go/SumRootToLeaf/SumRootToLeaf.go
package main

import (
	"fmt"
	"strconv"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sumNumbers(root *TreeNode) int {
	sum := 0
	traverse(root, "", &sum)
	return sum
}

func traverse(root *TreeNode, path string, sum *int) {
	if root == nil {
		return
	}

	path += fmt.Sprintf("%v", root.Val)
	if root.Left == nil && root.Right == nil {
		pathInt, _ := strconv.Atoi(path)
		*sum += pathInt
		return
	}

	traverse(root.Left, path, sum)
	traverse(root.Right, path, sum)
}

func main() {
	left := &TreeNode{Val: 2}
	right := &TreeNode{Val: 3}
	root := &TreeNode{Val: 1, Left: left, Right: right}
	fmt.Println(sumNumbers(root))
}
